clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -980,6 +998,49 @@ void handleCoordinatorReady() {
                 null;
         this.coordinatorSupportsBumpingEpoch = initProducerIdVersion != null &&
                 initProducerIdVersion.maxVersion() >= 3;
+
+        if (nodeApiVersions == null) return;
+
+        if (nodeApiVersions.supportedFeatures() != null) {
+            /*
+                To enable the transaction V2, it requires:
+                1. transaction.version max version >= 2
+                2. The ProduceRequest max version > ProducerRequest.LAST_BEFORE_TRANSACTION_V2_VERSION
+                3. The TxnOffsetCommitRequest max version > TxnOffsetCommitRequest.LAST_BEFORE_TRANSACTION_V2_VERSION
+            */
+            ApiVersion produceVersion = nodeApiVersions.apiVersion(ApiKeys.PRODUCE);
+            ApiVersion txnOffsetCommitVersion = nodeApiVersions.apiVersion(ApiKeys.TXN_OFFSET_COMMIT);
+            SupportedVersionRange transactionVersion = nodeApiVersions.supportedFeatures().get("transaction.version");
```

Comment:
This should be finalizedFeatures.

Reply:
Done. I am surprised that NodeApiVersion does not store the finalized features. Maybe I missed something.

Reply:
It probably wasn't needed until now?